Efficiency and bugfixes

mr-russ opened the pull request:
- Test extensions are loaded before starting.
- Enable (( in views to be processed.
- Enable bit field defaults to process.
- Use client buffered copy, so you don't need server privileges or the same computer for it to work.
- Use unbuffered mysql query and chunk sizes for copy, including individual load on error, so large chunks aren't missed, only individual rows.
- Add test files to allow faster checking that operations are working.

A 33Gb database was migrated in 54minutes from MariaDB 10 to PostgreSQL 9.6.    Without error checking and full tables, 25minutes can be acheived.

AnatolyUss commented on migration/FromMySqlToPostgreSql/FromMySqlToPostgreSql.php:
```diff
@@ -647,24 +659,30 @@ private function populateTableWorker(
-
-            if ($intRowsCnt != 0 && 0 == $intRetVal) {
-                $this->log("\t--Following MySQL query will return a data set, rejected by PostgreSQL:\n" . $sql . "\n");
+                if ($copiedCount % $intRowsInChunk) {
```
Could you clarify, what is this condition for?

AnatolyUss commented on migration/FromMySqlToPostgreSql/FromMySqlToPostgreSql.php:
```diff
@@ -647,24 +659,30 @@ private function populateTableWorker(
+                if ($copiedCount % $intRowsInChunk) {
+                    // Attempt copy, if it failes, do one at a time to ensure we find all the errors.
+                    // If the data is valid, we perform fast, otherwise we ensure correctness at the cost of speed.
+
```
In my humble experience, in most cases when copy fails, it is not due to some error in one single record, so if the script will perform one insertion per one record, likely it will result in
cascading failures. It will be very slow process.  
Even if the problem is only in one single record, which is very unlikely, inserting each record in a separate transaction is very slow.

AnatolyUss commented on migration/FromMySqlToPostgreSql/FromMySqlToPostgreSql.php:
```diff
@@ -911,6 +911,9 @@ private function processDefault($strTableName, array $arrColumns)
 
                 if (isset($arrSqlReservedValues[$arrColumn['Default']])) {
                     $sql .= $arrSqlReservedValues[$arrColumn['Default']] . ';';
+                } else if (substr($arrColumn['Type'], 0, 3) == 'bit' && substr($arrColumn['Default'], 0, 2) == "b'") {
```
Please, use only identical equality checks (<code>===</code>).

AnatolyUss commented on tests/test_config.json:
```diff
@@ -0,0 +1,41 @@
+{
```
In general, it is very good idea to add tests!  
But if you do so, then add a new section to <code>README.md</code>

mr-russ replied:
Section has now been added to the README.

AnatolyUss commented:
"Without error checking and full tables" - please, explain what do you mean?

mr-russ commented:
What I mean by no error checking is;

This change batches the results in the existing configuration batch size.  When a batch fails, it will copy each row into the target individually to ensure that only bad data is excluded and not whole chunks.  This was the case previously.

If I remove batches all together and stream the mysql input data directly into pg_put_line() as a copy operation, I can move very large tables very quickly.  However if there is any bad data in the table at all, the target table will end up empty.  So I opted for getting most data migrated over the possible performance that could be seen under best case scenario.

Even with the current changes, the performance is significantly better than the current published results and allows the script to run on a server that isn't the target database server.  Very useful if you are using a cloud provide that only exposes an endpoint.

I will look at the other issues you raised soon.

mr-russ commented:
I've addresses all the concerns raised as part of the code review.  I've also update the performance section as the new code is about 20x faster, so the old results don't say anything that would give a useful guide to somebody looking to use this tool.

mr-russ commented:
In relation to your copy failing issues.  My 15 years experience is exactly the opposite with PostgreSQL.  It's always a single row data issue.  What you would have seen with the migration before, and @chewi identifies this really well, CSV is a horrible import format.  It's never escaped properly and PHP doesn't handle it properly.
The most common issue I have with data migration is invalid MySQL dates, integer overflow issues unsigned MySQL ints not fitting in signed PostgreSQL fields or encoding problems that can't be imported.

I would be interested in seeing a set of data that has mass copy problems between the two databases.  With the 33Gb database I migrationed, I had a single row failure because of the date 2000-01-01.  Which is not a real date.

Thanks for your review and I look forward to any further comments and getting this merged.

Russell
